### PR TITLE
feat: add ButtonGroup component

### PR DESCRIPTION
**Pull request by snitin315**
## Description

Adds new `ButtonGroup` component to Blade.

## Changes

- API - https://github.com/razorpay/blade/pull/2061
- Added scale-down animation to the button when pressed.
- Added ButtonGroup.
- Tests - https://github.com/razorpay/blade/pull/2068
- [FC - 2729](https://razorpay.atlassian.net/browse/FC-2729?atlOrigin=eyJpIjoiZDg1ZTZiYmM2OWQ3NDQ5OGExZjExNWZmZjM5YWVlMWYiLCJwIjoiaiJ9)

## Component Checklist

- [ ] Update Component Status Page
- [x] Perform Manual Testing in Other Browsers
- [x] Add KitchenSink Story
- [x] Add Interaction Tests (if applicable)
- [x] Add changeset

### CODE REVIEW

File: packages/blade/src/components/ButtonGroup/ButtonGroup.web.tsx [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        {...metaAttribute({ name: MetaConstants.ButtonGroup, testID })}
+        {...getStyledProps(styledProps)}
+        role="group"
+      >
+        {React.Children.map(children, (child, index) => {
+          if (__DEV__) {
+            // throw error if child is not a button or dropdown with button trigger
+            /* eslint-disable no-restricted-properties */
+            if (
+              !isValidAllowedChildren(child, 'Button') &&
```
Comment (anuraghazra) [inline packages/blade/src/components/ButtonGroup/ButtonGroup.web.tsx:88]: Can we use `componentIds.Button`?

**Review by anuraghazra — DISMISSED** [review]
Build is failing, unable to see storybook preview.

**Comment by snitin315** [comment]
@anuraghazra [Build is successful,](https://github.com/razorpay/blade/actions/runs/8262733379/job/22602699087?pr=2064) for some reason it's not coming in CI checks as usual. 

Preview Link - https://61c19ee8d3d282003ac1d81c-mdcfdyzkei.chromatic.com/

File: packages/blade/src/components/Button/BaseButton/BaseButton.tsx [hunk 2]
```diff
@@ -412,16 +416,46 @@ const _BaseButton: React.ForwardRefRenderFunction<BladeElementRef, BaseButtonPro
     buttonTypographyTokens: buttonTypography,
     children: childrenString,
     isDisabled: disabled,
-    size,
-    variant,
+    size: buttonGroupProps.size ?? size,
+    variant: buttonGroupProps.variant ?? variant,
     theme,
-    color,
+    color: buttonGroupProps.color ?? color,
```
Comment (saurabhdaware) [inline packages/blade/src/components/Button/BaseButton/BaseButton.tsx:422]: expected to give ButtonGroup a priority over individual Button?

File: packages/blade/src/components/Button/BaseButton/BaseButton.tsx [hunk 3]
```diff
@@ -452,85 +486,106 @@ const _BaseButton: React.ForwardRefRenderFunction<BladeElementRef, BaseButtonPro
...
       onMouseMove={onMouseMove}
       onPointerDown={onPointerDown}
       onPointerEnter={onPointerEnter}
-      onKeyDown={onKeyDown}
-      onTouchStart={onTouchStart}
-      onTouchEnd={onTouchEnd}
+      // Setting type for web fails it on native typecheck and vice versa
+      onKeyDown={(event: any) => {
+        handleKeyboardPressedIn(event);
+        if (onKeyDown) onKeyDown(event);
```
Comment (saurabhdaware) [inline packages/blade/src/components/Button/BaseButton/BaseButton.tsx:500]:
```suggestion
        onKeyDown?.(event);
```

Comment (saurabhdaware) [reply packages/blade/src/components/Button/BaseButton/BaseButton.tsx:500]: same thing with other event handlers

File: packages/blade/src/components/ButtonGroup/ButtonGroup.native.tsx [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import type { ButtonGroupProps } from './types';
+import { throwBladeError } from '~utils/logger';
+
+const ButtonGroup = (_props: ButtonGroupProps): React.ReactElement => {
```
Comment (saurabhdaware) [inline packages/blade/src/components/ButtonGroup/ButtonGroup.native.tsx:4]: are there any specific changes we need to make ButtonGroup work on react native? Like if we just export ButtonGroup from here, is it working on react native? cause I didn't see much of the web-specific code.

File: packages/blade/src/components/ButtonGroup/StyledButtonGroup.tsx [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+  ({ theme, isDisabled, variant = 'primary', color, isFullWidth }) => {
+    return {
+      display: 'flex',
+      width: isFullWidth ? '100%' : 'fit-content',
+      borderRadius: makeBorderSize(theme.border.radius.medium),
+      borderWidth: makeBorderSize(theme.border.width.thin),
+      borderColor: getIn(
+        theme.colors,
+        getBackgroundColorToken({
+          property: variant === 'tertiary' ? 'background' : 'border',
```
Comment (saurabhdaware) [inline packages/blade/src/components/ButtonGroup/StyledButtonGroup.tsx:18]: why is tertiary button using `background` token for border? can you check with designer once?

File: packages/blade/src/components/ButtonGroup/ButtonGroup.web.tsx [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+              )}
+            </>
+          );
+        })}
+      </StyledButtonGroup>
+    </ButtonGroupProvider>
+  );
+};
+
+const ButtonGroup = assignWithoutSideEffects(_ButtonGroup, {
```
Comment (saurabhdaware) [inline packages/blade/src/components/ButtonGroup/ButtonGroup.web.tsx:118]: add jsdoc with basic usage?

File: packages/blade/src/components/Button/BaseButton/BaseButton.tsx (hunk 2 above)
Comment (snitin315) [reply packages/blade/src/components/Button/BaseButton/BaseButton.tsx:422]: Yes, all buttons should have the same color in a button group.

File: packages/blade/src/components/ButtonGroup/ButtonGroup.native.tsx (hunk 4 above)
Comment (snitin315) [reply packages/blade/src/components/ButtonGroup/ButtonGroup.native.tsx:4]: Yes, I believe there will be some CSS modifications. Currently if we just export this file it's visually broken.

File: packages/blade/src/components/ButtonGroup/StyledButtonGroup.tsx (hunk 5 above)
Comment (snitin315) [reply packages/blade/src/components/ButtonGroup/StyledButtonGroup.tsx]: we need it for secondary variant only, fixed.

File: packages/blade/src/components/ButtonGroup/ButtonGroup.web.tsx (hunk 6 above)
Comment (snitin315) [reply packages/blade/src/components/ButtonGroup/ButtonGroup.web.tsx:144]: Done

File: packages/blade/src/components/ButtonGroup/ButtonGroup.web.tsx (hunk 1 above)
Comment (anuraghazra) [reply packages/blade/src/components/ButtonGroup/ButtonGroup.web.tsx:88]:
```suggestion
              !isValidAllowedChildren(child, componentIds.Button) &&
```

File: packages/blade/src/components/ButtonGroup/index.ts [hunk 7]
```diff
@@ -0,0 +1 @@
+export * from './ButtonGroup';
```
Comment (anuraghazra) [inline packages/blade/src/components/ButtonGroup/index.ts:1]: export the types too.

**Review by anuraghazra — APPROVED** [review]

**Review by saurabhdaware — APPROVED** [review]

File: packages/blade/src/components/ButtonGroup/index.ts (hunk 7 above)
Comment (snitin315) [reply packages/blade/src/components/ButtonGroup/index.ts:1]: It's already exported.

https://github.com/razorpay/blade/blob/eb5283b3f0b36cd0804d43ec029be1f95c7eb2c1/packages/blade/src/components/ButtonGroup/ButtonGroup.web.tsx#L149-L150

**Pull request merged**